### PR TITLE
QRegEx to QRegularExpression

### PR DESCRIPTION
**Pull request by roya0045**
## Description
 Supersedes #41540 

Should contain only changes that can be merged currently.

Wildcards  and anchoredpattern are in another branch.

### CODE REVIEW

File: python/core/auto_generated/qgsapplication.sip.in [hunk 1]
```diff
@@ -378,7 +378,7 @@ colors are specified then the ``size`` argument also must be set.
 Returns the path to user's style.
 %End
 
-    static QRegExp shortNameRegExp();
+    static QRegularExpression shortNameRegExp();
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsapplication.sip.in:381]: You'll need to revert anything which changes the Python bindings -- because these are part of the stable API and QRegularExpression isn't api compatible with QRegEx, we risk breaking existing python plugins and scripts if we alter the types now.

File: python/plugins/db_manager/db_plugins/data_model.py [hunk 2]
```diff
@@ -290,11 +290,11 @@ def getObject(self, row):
...
         fld.name = self.data(self.index(row, 0)) or ""
         typestr = self.data(self.index(row, 1)) or ""
-        regex = QRegExp("([^\\(]+)\\(([^\\)]+)\\)")
-        startpos = regex.indexIn(typestr)
-        if startpos >= 0:
-            fld.dataType = regex.cap(1).strip()
-            fld.modifier = regex.cap(2).strip()
+        regex = QRegularExpression("([^\\(]+)\\(([^\\)]+)\\)")
+        matches = regex.match(typestr)
+        if matches.capturedStart() >= 0:
```
Comment (nyalldawson) [inline python/plugins/db_manager/db_plugins/data_model.py:295]:
```suggestion
        if matches.hasMatch():
```

File: python/plugins/db_manager/db_plugins/postgis/connector.py [hunk 3]
```diff
@@ -851,9 +851,10 @@ def getSpatialRefInfo(self, srid):
 
         srtext = sr[0]
         # try to extract just SR name (should be quoted in double quotes)
-        regex = QRegExp('"([^"]+)"')
-        if regex.indexIn(srtext) > -1:
-            srtext = regex.cap(1)
+        regex = QRegularExpression('"([^"]+)"')
+        matches = regex.match(srtext)
+        if matches.capturedStart() > -1:
```
Comment (nyalldawson) [inline python/plugins/db_manager/db_plugins/postgis/connector.py:856]:
```suggestion
        if matches.hasMatch():
```

File: python/plugins/db_manager/db_plugins/data_model.py [hunk 4]
```diff
@@ -290,11 +290,11 @@ def getObject(self, row):
         fld = val if val is not None else self._getNewObject()
         fld.name = self.data(self.index(row, 0)) or ""
         typestr = self.data(self.index(row, 1)) or ""
-        regex = QRegExp("([^\\(]+)\\(([^\\)]+)\\)")
-        startpos = regex.indexIn(typestr)
-        if startpos >= 0:
-            fld.dataType = regex.cap(1).strip()
-            fld.modifier = regex.cap(2).strip()
+        regex = QRegularExpression("([^\\(]+)\\(([^\\)]+)\\)")
```
Comment (nyalldawson) [inline python/plugins/db_manager/db_plugins/data_model.py:293]:
```suggestion
        regex = QRegularExpression(r"([^\(]+)\(([^\)]+)\)")
```

File: python/plugins/db_manager/db_plugins/postgis/plugin.py [hunk 5]
```diff
@@ -394,10 +394,10 @@ def __init__(self, row, table):
 
         # get modifier (e.g. "precision,scale") from formatted type string
         trimmedTypeStr = typeStr.strip()
-        regex = QRegExp("\\((.+)\\)$")
-        startpos = regex.indexIn(trimmedTypeStr)
-        if startpos >= 0:
-            self.modifier = regex.cap(1).strip()
+        regex = QRegularExpression("\\((.+)\\)$")
```
Comment (nyalldawson) [inline python/plugins/db_manager/db_plugins/postgis/plugin.py:397]:
```suggestion
        regex = QRegularExpression(r"\((.+)\)$")
```

File: src/app/options/qgsoptions.cpp [hunk 6]
```diff
@@ -2781,8 +2781,8 @@ void QgsOptions::updateActionsForCurrentColorScheme( QgsColorScheme *scheme )
 void QgsOptions::scaleItemChanged( QListWidgetItem *changedScaleItem )
 {
   // Check if the new value is valid, restore the old value if not.
-  QRegExp regExp( "1:0*[1-9]\\d*" );
-  if ( regExp.exactMatch( changedScaleItem->text() ) )
+  QRegularExpression regExp( "1:0*[1-9]\\d*" );
```
Comment (nyalldawson) [inline src/app/options/qgsoptions.cpp:2784]: need anchoredPattern

File: python/plugins/db_manager/db_plugins/postgis/plugin.py [hunk 7]
```diff
@@ -394,10 +394,10 @@ def __init__(self, row, table):
...
+        matches = regex.match(trimmedTypeStr)
+        if matches.captutredStart() >= 0:
```
Comment (nyalldawson) [inline python/plugins/db_manager/db_plugins/postgis/plugin.py:399]:
```suggestion
        if matches.hasMatch():
```

File: src/app/qgisapp.cpp [hunk 8]
```diff
@@ -5997,13 +6000,14 @@ QList< QgsMapLayer * > QgisApp::askUserForGDALSublayers( QgsRasterLayer *layer )
     for ( const QgsSublayersDialog::LayerDefinition &def : constSelection )
     {
       int i = def.layerId;
-
       const QStringList parts = sublayers[i].split( QgsDataProvider::sublayerSeparator() );
       const QString path = parts[0];
       QString name = path;
-      if ( rx.indexIn( name ) != -1 )
+      matches = rx.match( sublayers[i] );
+      if ( matches.capturedStart() != -1 )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:6007]:
```suggestion
      if ( matches.hasMatch() )
```

File: src/app/qgisapp.cpp [hunk 9]
```diff
@@ -5997,13 +6000,14 @@ QList< QgsMapLayer * > QgisApp::askUserForGDALSublayers( QgsRasterLayer *layer )
...
       {
-        const QString uri = rx.cap( 1 );
+        uri = matches.captured( 1 );
+        name = sublayers[i];
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:6010]: why this change?

File: src/app/qgisapp.cpp [hunk 10]
```diff
@@ -12988,8 +12992,8 @@ QgsVectorLayer *QgisApp::addVectorLayerPrivate( const QString &vectorLayerPath,
 
   // if the layer needs authentication, ensure the master password is set
   bool authok = true;
-  QRegExp rx( "authcfg=([a-z]|[A-Z]|[0-9]){7}" );
-  if ( rx.indexIn( vectorLayerPath ) != -1 )
+  QRegularExpression rx( "authcfg=([a-z]|[A-Z]|[0-9]){7}" );
+  if ( rx.match( vectorLayerPath ).capturedStart() != -1 )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:12996]:
```suggestion
  if ( rx.match( vectorLayerPath ).hasMatch() )
```

File: src/app/qgsprojectproperties.cpp [hunk 11]
```diff
@@ -2553,8 +2553,8 @@ void QgsProjectProperties::addScaleToScaleList( QListWidgetItem *newItem )
 void QgsProjectProperties::scaleItemChanged( QListWidgetItem *changedScaleItem )
 {
   // Check if the new value is valid, restore the old value if not.
-  QRegExp regExp( "1:0*[1-9]\\d*" );
-  if ( regExp.exactMatch( changedScaleItem->text() ) )
+  QRegularExpression regExp( "1:0*[1-9]\\d*" );
```
Comment (nyalldawson) [inline src/app/qgsprojectproperties.cpp:2556]: Needs anchoredPattern

Comment (nyalldawson) [reply src/app/qgsprojectproperties.cpp:2556]: please check other occurrences of exactMatch have all been ported to anchoredPattern, and replace other capturedStart(...) >= 0 checks with .hasMatch().

Ping me when this is done and I'll continue the review :+1:

**Review by nyalldawson — CHANGES_REQUESTED** [review]

File: src/app/qgsprojectproperties.cpp (hunk 11 above)
Comment (roya0045) [reply src/app/qgsprojectproperties.cpp:2556]: Most of those should be in the other branch, as we updated dependencies, do you want me to merge the two, I was not sure how to best handle this considering the recent changes.

File: python/core/auto_generated/qgsapplication.sip.in (hunk 1 above)
Comment (roya0045) [reply python/core/auto_generated/qgsapplication.sip.in:381]: Sure thing!

File: src/app/qgsprojectproperties.cpp (hunk 11 above)
Comment (nyalldawson) [reply src/app/qgsprojectproperties.cpp:2556]: let's just get this one merged and then see what changes remain in the other branch?

File: python/core/auto_generated/qgsapplication.sip.in (hunk 1 above)
Comment (roya0045) [reply python/core/auto_generated/qgsapplication.sip.in:381]: I have a question regarding the best approach to this, for functions that require regex input, should I duplicate them and convert the regex to regulareexpression to call the new version or should I not do any conversion and leave untouched anything that would need a regexp in the API ( this might be a lot of stuff to revert).

File: python/plugins/db_manager/db_plugins/data_model.py (hunk 2 above)
Comment (roya0045) [reply python/plugins/db_manager/db_plugins/data_model.py]: Changed them all.

File: python/plugins/db_manager/db_plugins/data_model.py (hunk 4 above)
Comment (roya0045) [reply python/plugins/db_manager/db_plugins/data_model.py]: Done thanks!

File: python/core/auto_generated/qgsapplication.sip.in (hunk 1 above)
Comment (roya0045) [reply python/core/auto_generated/qgsapplication.sip.in]: Api should be safe now.

File: src/app/qgisapp.cpp (hunk 9 above)
Comment (roya0045) [reply src/app/qgisapp.cpp]: can't recall, reverted now.

**Comment by roya0045** [comment]
closing temporarily to investigate the test results.
